homeassistant/components/yeelight/__init__.py

```diff
@@ -1 +1,357 @@
+            cv.time_period,
+        vol.Optional(CONF_CUSTOM_EFFECTS): [{
+            vol.Required(CONF_NAME): cv.string,
+            vol.Required(CONF_FLOW_PARAMS): YEELIGHT_FLOW_TRANSITION_SCHEMA
+        }]
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+YEELIGHT_SERVICE_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+NIGHTLIGHT_SUPPORTED_MODELS = [
+    "ceiling3",
+    'ceiling4'
+]
+
+UPDATE_REQUEST_PROPERTIES = [
+    "power",
+    "bright",
+    "ct",
+    "rgb",
+    "hue",
+    "sat",
+    "color_mode",
+    "flowing",
+    "music_on",
+    "nl_br",
+    "active_mode",
+]
+
+
+def _transitions_config_parser(transitions):
+    """Parse transitions config into initialized objects."""
+    import yeelight
+
+    transition_objects = []
+    for transition_config in transitions:
+        transition, params = list(transition_config.items())[0]
+        transition_objects.append(getattr(yeelight, transition)(*params))
+
+    return transition_objects
+
+
+def _parse_custom_effects(effects_config):
+    import yeelight
+
+    effects = {}
+    for config in effects_config:
+        params = config[CONF_FLOW_PARAMS]
+        action = yeelight.Flow.actions[params[ATTR_ACTION]]
+        transitions = _transitions_config_parser(
+            params[ATTR_TRANSITIONS])
+
+        effects[config[CONF_NAME]] = {
+            ATTR_COUNT: params[ATTR_COUNT],
+            ATTR_ACTION: action,
+            ATTR_TRANSITIONS: transitions
+        }
+
+    return effects
+
+
+def setup(hass, config):
+    """Set up the Yeelight bulbs."""
+    from yeelight.enums import PowerMode
+
+    conf = config[DOMAIN]
+    yeelight_data = hass.data[DATA_YEELIGHT] = {
+        CONF_DEVICES: {},
+        CONF_LIGHTS: {},
+    }
+
+    def device_discovered(service, info):
+        _LOGGER.debug("Adding autodetected %s", info['hostname'])
+
+        device_type = info['device_type']
+
+        name = "yeelight_%s_%s" % (device_type,
+                                   info['properties']['mac'])
+        ipaddr = info[CONF_HOST]
+        device_config = DEVICE_SCHEMA({
+            CONF_NAME: name,
+            CONF_MODEL: device_type
+        })
+
+        _setup_device(hass, config, ipaddr, device_config)
+
+    discovery.listen(hass, SERVICE_YEELIGHT, device_discovered)
+
+    def async_update(event):
+        for device in yeelight_data[CONF_DEVICES].values():
+            device.update()
+
+    async_track_time_interval(
+        hass, async_update, conf[CONF_SCAN_INTERVAL]
+    )
+
+    def service_handler(service):
+        """Dispatch service calls to target entities."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+
+        entity_ids = extract_entity_ids(hass, service)
+        target_devices = [dev.device for dev in
+                          yeelight_data[CONF_LIGHTS].values()
+                          if dev.entity_id in entity_ids]
```

Comment:
We don't need to access the entity here. It's enough to check for the correct entity_ids. See comment below about storing entities vs entity_ids.